backend-tests/tests/test_devauth_v2.py

```diff
@@ -0,0 +1,277 @@
+def devices(clean_migrated_mongo):
+    devauthd = ApiClient(deviceauth_v1.URL_DEVICES)
+
+    devices = []
+
+    for _ in range(5):
+        d = create_random_device()
+        devices.append(d)
+
+    yield devices
+
+@pytest.yield_fixture(scope="function")
+def tenants_users(clean_migrated_mongo_mt):
+    cli = CliTenantadm()
+    api = ApiClient(tenantadm.URL_INTERNAL)
+
+    names = ['tenant1', 'tenant2']
+    tenants=[]
+
+    for n in names:
+        tenants.append(create_tenant(n))
+
+    for t in tenants:
+        for i in range(2):
+            user = create_tenant_user(i, t)
+            t.users.append(user)
+
+    yield tenants
+
+@pytest.yield_fixture(scope="function")
+def tenants_users_devices(clean_migrated_mongo_mt, tenants_users):
+    for t in tenants_users:
+        for _ in range(5):
+            dev = create_random_device(t.tenant_token)
+            t.devices.append(dev)
+
+    yield tenants_users
+
+class TestPreauthBase:
+    def do_test_ok(self, user, tenant_token=''):
+        useradmm = ApiClient(useradm.URL_MGMT)
+        devauthm = ApiClient(deviceauth_v2.URL_MGMT)
+        devauthd = ApiClient(deviceauth_v1.URL_DEVICES)
+
+        # log in user
+        r = useradmm.call('POST',
+                          useradm.URL_LOGIN,
+                          auth=(user.name, user.pwd))
+        assert r.status_code == 200
+
+        utoken = r.text
+
+        # preauth device
+        priv, pub = util.crypto.rsa_get_keypair()
+        id_data = {'mac': 'pretenditsamac'}
+        body = deviceauth_v2.preauth_req(
+                    id_data,
+                    pub)
+        r = devauthm.with_auth(utoken).call('POST',
+                                            deviceauth_v2.URL_DEVICES,
+                                            body)
+        assert r.status_code == 201
+
+        # device appears in device list
+        r = devauthm.with_auth(utoken).call('GET',
+                                            deviceauth_v2.URL_DEVICES)
+        assert r.status_code == 200
+        api_devs = r.json()
+
+        assert len(api_devs) == 1
+        api_dev = api_devs[0]
+
+        assert api_dev['status'] == 'preauthorized'
+        assert api_dev['identity_data'] == id_data
+        assert len(api_dev['auth_sets']) == 1
+        aset = api_dev['auth_sets'][0]
+
+        assert aset['identity_data'] == id_data
+        assert util.crypto.rsa_compare_keys(aset['pubkey'], pub)
+        assert aset['status'] == 'preauthorized'
+
+        # actual device can obtain auth token
+        body, sighdr = deviceauth_v1.auth_req(id_data,
+                                              pub,
+                                              priv,
+                                              tenant_token)
+
+        r = devauthd.call('POST',
+                          deviceauth_v1.URL_AUTH_REQS,
+                          body,
+                          headers=sighdr)
+
+        assert r.status_code == 200
```

Comment:
what about checking if the aset and device status was updated after device request?

Reply:
fixed (stacked commit)